File: matiasmasso/Mat.NET/DTO/DTOCcb.cs
using System;

namespace DTO
{
    public class DTOCcb : DTOBaseGuid
    {
        public DTOCca Cca { get; set; }
        public DTOContact Contact { get; set; }
        public DTOPgcCta Cta { get; set; }
        public DTOAmt Amt { get; set; }
        public DhEnum Dh { get; set; }
        public int Lin { get; set; }
        public DTOPnd Pnd { get; set; }

        public enum DhEnum
        {
            notSet,
            debe,
            haber
        }

        public DTOCcb() : base()
        {
        }

        public DTOCcb(Guid oGuid) : base(oGuid)
        {
        }

        public static DTOCcb Factory(DTOCca oCca, DTOAmt oAmt, DTOPgcCta oCta, DTOContact oContact, DTOCcb.DhEnum oDh, DTOPnd oPnd = null/* TODO Change to default(_) if this is not a reference type */)
        {
            DTOCcb retval = new DTOCcb();
            {
                var withBlock = retval;
                withBlock.Cca = oCca;
                withBlock.Amt = oAmt;
                withBlock.Cta = oCta;
                withBlock.Contact = oContact;
                withBlock.Dh = oDh;
                withBlock.Pnd = oPnd;
            }
            return retval;
        }

        public static DTOAmt Debit(DTOCcb oCcb)
        {
            DTOAmt retval = null/* TODO Change to default(_) if this is not a reference type */;
            if (oCcb.Dh == DTOCcb.DhEnum.debe)
                retval = oCcb.Amt;
            return retval;
        }

        public static DTOAmt Credit(DTOCcb oCcb)
        {
            DTOAmt retval = null/* TODO Change to default(_) if this is not a reference type */;
            if (oCcb.Dh == DTOCcb.DhEnum.haber)
                retval = oCcb.Amt;
            return retval;
        }

        public static void Arrastra(DTOCcb oCcb, ref DTOAmt oDeure, ref DTOAmt oHaver, ref DTOAmt oSaldo)
        {
            if (oCcb.Dh == DTOCcb.DhEnum.debe)
                oDeure.Add(oCcb.Amt);
            else
                oHaver.Add(oCcb.Amt);

            if ((int)(oCcb.Cta.Act) == (int)(oCcb.Dh))
                oSaldo.Add(oCcb.Amt);
            else
                oSaldo.Substract(oCcb.Amt);
        }
    }
}

File: matiasmasso/Mat.NET/DT2/DTOCcb.vb
Public Class DTOCcb
    Inherits DTOBaseGuid
    Property cca As DTOCca
    Property contact As DTOContact
    Property cta As DTOPgcCta
    Property amt As DTOAmt
    Property dh As DhEnum
    Property lin As Integer
    Property pnd As DTOPnd

    Public Enum DhEnum
        notSet
        debe
        haber
    End Enum

    Public Sub New()
        MyBase.New()
    End Sub

    Public Sub New(oGuid As Guid)
        MyBase.New(oGuid)
    End Sub

    Shared Function Factory(oCca As DTOCca, oAmt As DTOAmt, oCta As DTOPgcCta, oContact As DTOContact, oDh As DTOCcb.DhEnum, Optional oPnd As DTOPnd = Nothing) As DTOCcb
        Dim retval As New DTOCcb
        With retval
            .Cca = oCca
            .Amt = oAmt
            .Cta = oCta
            .Contact = oContact
            .Dh = oDh
            .Pnd = oPnd
        End With
        Return retval
    End Function

    Shared Function Debit(oCcb As DTOCcb) As DTOAmt
        Dim retval As DTOAmt = Nothing
        If oCcb.Dh = DTOCcb.DhEnum.Debe Then
            retval = oCcb.Amt
        End If
        Return retval
    End Function

    Shared Function Credit(oCcb As DTOCcb) As DTOAmt
        Dim retval As DTOAmt = Nothing
        If oCcb.Dh = DTOCcb.DhEnum.Haber Then
            retval = oCcb.Amt
        End If
        Return retval
    End Function

    Shared Sub Arrastra(oCcb As DTOCcb, ByRef oDeure As DTOAmt, ByRef oHaver As DTOAmt, ByRef oSaldo As DTOAmt)
        If oCcb.Dh = DTOCcb.DhEnum.Debe Then
            oDeure.Add(oCcb.Amt)
        Else
            oHaver.Add(oCcb.Amt)
        End If

        If oCcb.Cta.Act = oCcb.Dh Then
            oSaldo.Add(oCcb.Amt)
        Else
            oSaldo.Substract(oCcb.Amt)
        End If
    End Sub


End Class
